### PR TITLE
jax.scipy.special: fix gradient for xlogy & xlog1py

### LINKED ISSUE #15598: Incorrect gradient for `jax.scipy.stats.beta.logpdf`
### Description

Hi all. During some testing for an inference problem, I noticed that the gradients computed for `jax.scipy.stats.beta.logpdf` were incorrect when using alpha/beta parameters at 1. Please find a MWE below.

```python
import jax
import jax.numpy as jnp
import jax.scipy.special as jaxspec
import jax.scipy.stats as jaxstats

from jax import Array
from jax.config import config
from jax.scipy.special import gammaln
from jax.typing import ArrayLike
config.update("jax_enable_x64", True)


# custom loglikelihood
def jax_loglik(params: ArrayLike, p: ArrayLike) -> Array:
    k, n = params
    return (
        (k - 1) * jnp.sum(jnp.log(p))
        + (n - 1) * jnp.sum(jnp.log1p(-p))
        - len(p) * (gammaln(k) + gammaln(n) - gammaln(k + n))
    )

# log likelihood using direct logpdf sum
def jax_logpdf(params: ArrayLike, p: ArrayLike) -> Array:
    k, n = params
    return jnp.sum(jaxstats.beta.logpdf(p, k, n))

# fake data
x = 0.5 * jnp.ones(2)

# problematic parameters
bad_params = jnp.ones(2)

# non-problematic parameters
fine_params = 2 * jnp.ones(2)

for name, param in [("bad param", bad_params), ("fine param", fine_params)]:
    print(f"Evaluating functions using {name}")

    # eval each log likelihood definition using alpha = 1., beta = 1., x1 = 0.5, x2 = 0.5
    print(f"Bespoke ll = {jax_loglik(param, x)}")
    print(f"JAX beta.logpdf = {jax_logpdf(param, x)}")
    
    # eval gradient of log likelihood wrt alpha, beta, at alpha = beta = 1.
    print(f"Grad bespoke ll = {jax.grad(jax_loglik)(param, x)}")
    print(f"Grad JAX beta.logpdf {jax.grad(jax_logpdf)(param, x)}")
    print("")
```

The above code generates the following output, 
```
Evaluating functions using bad param
Bespoke ll = -0.0
JAX beta.logpdf = 0.0
Grad bespoke ll = [0.61370564 0.61370564]
Grad JAX beta.logpdf [2. 2.]

Evaluating functions using fine param
Bespoke ll = 0.8109302162163301
JAX beta.logpdf = 0.8109302162163301
Grad bespoke ll = [0.28037231 0.28037231]
Grad JAX beta.logpdf [0.28037231 0.28037231]
```

### What jax/jaxlib version are you using?

jax v0.4.8

### Which accelerator(s) are you using?

CPU

### Additional system info

Mac M1

### NVIDIA GPU info

_No response_

### PR DESCRIPTION
**Pull request by jakevdp**
Fixes #15598

### CODE REVIEW

**Review by sharadmv — APPROVED** [review]

**Comment by mattjj** [comment]
Can we use `jtu.check_grads` to check higher order AD too? (At 0 and elsewhere.)

**Comment by mattjj** [comment]
Do those test at 0 though? If they did they already would have caught this bug, right?

**Comment by jakevdp** [comment]
Oh I see, but to be clear we don't have higher-order coverage for *any* `jax.scipy.special` functions: https://github.com/google/jax/blob/933d695170246efaba2f949b774235e407f1ea08/tests/lax_scipy_special_functions_test.py#L176

**Comment by mattjj** [comment]
Ah something simpler than that, just take the current tests and add one new line to each, like `jtu.check_grads(lsp_special.xlogy, (0.0, 3.0), order=2)` and `jtu.check_grads(lsp_special.xlog1py, (0.0, 3.0), order=2)`.

**Comment by mattjj** [comment]
Ah that may not work because `check_jvp` and `check_vjp` may choose a negative step direction: https://github.com/google/jax/blob/933d695170246efaba2f949b774235e407f1ea08/jax/_src/public_test_util.py#L196

Nevermind, we should improve those utilities to handle this kind of case, but not block this PR on it.
